### PR TITLE
Goals

### PR DESCRIPTION
**Pull request by chrishaid**
@almartin82 :  Here's the goals/projections architecture and code I promised. Take a good look through the code. But pay attention to the `add_accelerated_goals` function, which is the workhorse (albeit, a very simple one). that function essentially calls a "goals" function which creates a "goals" object (nothing formal here, though there are some ensureR promises for validation). That object is a list with three elements: 
1. a `data.frame` named `goals` with the goals and any informative data (like studentids, testids, etc).  `goals` **must have** `accel_growth` and `met_accel_growth` columns.
2. a character vector `join_by_fields` announcing which columns for an inner join with the `growth_df` in a `mapvizieR` object, should you ever care to do such a thing (which you will!).
3. a `slot_name` variable, which provides a means of naming the slot which this list will be attached to on a `mapvizieR` objects `goals` attribute (this attribute is new and now implemented).

To reiterate: `add_accelerated_growth` runs a "goals" function which generates this "goals proto-object" and then slides that object into a new slot under the `goals` attribute of a `mapvizieR` object.  If you set the `update_growth_df` parameter to `TRUE`, then  `add_accelerated_growth` will do the `dplyr::inner_join` on the "goals" object with `growth_df` attribute of the `mapvizieR` object, effectively updating that object by added an accelerated growth column and met accelerated growth column.  The whole shebang returns a new mapvizier object with the `goals` attribute added. 

The only "goals" function right now is `goal_kipp_tiered()` which generates a "goals object" as defined above.  

I added this functionality to the `mapvizieR` constructor function (i.e. there is a `goals` slot with KIPP tiered growth and the `growth_df` has the two extra columns. 

This should provide a really flexible architecture of dealing with accelerated goals, predictions, and simulations and handling graphing from the object.  I'm pretty psyched by this one.

### CODE REVIEW

File: R/mapvizieR_object.R [hunk 1]
```diff
@@ -49,6 +49,17 @@ mapvizieR.default <- function(raw_cdf, raw_roster) {
   )
   
   class(mapviz) <- "mapvizieR"
+
```
Comment (almartin82) [inline R/mapvizieR_object.R:52]: any thoughts on why the object becomes class `mapvizieR` before the accelerated goals?

Comment (chrishaid) [inline R/mapvizieR_object.R:52]: See the comment, but it simplified writing the add_accelerated_growth function, which requires a mapvizieR object to run.  This allows us to use the function inside _and_ outside of the constructor function while keeping it nice, simple, maintainable, etc.

**Comment by chrishaid** [comment]
@almartin82 For posterity, **the current state of mapvizieR object** architecture:

```
mapvizieR_object (a mapvizieR object which is essentially a list)
|----cdf (a data.frame)
|----roster (a data.frame)
|----growth_df (another data.frame)
|----goals (a list with "custom" named slots)
      |----kipp_tiered_growth (a list, this is an example, requires three elements below)
            |----goals (a data frame, requires columns `accel_growth`, `met_accel_growth`)
            |----join_by_fields (a character vector, naming columns needed to join goals back to growth_df)
            |----slot_name (a 1 element character vector of the name for the slot.  In this case populated with "kipp_tiered_growth"
      |----rutgers_ready_growth (another example of a "growth object")
            |----goals
            |----join_by_fields 
            |----slot_name
      .
      . (so on and so forth)
      .
      |----some_other_growth_simulation
            |----goals
            |----join_by_fields
            |----slot_name



```

**Comment by chrishaid** [comment]
The biggest problem with the above architecture is the freely named slots after `mapvizier$goals`.  This may be providing too much flexibility at the detriment of our knowing the name structure. 

@almartin82 is proposing something simpler:

```
mapvizieR_object 
|----cdf (a data.frame)
|----roster (a data.frame)
|----growth_df (another data.frame)
|----goals (a data.frame with required fields: accel_growth, met_accel_growth, and growth_type)
|----simulated_goals (a list, maybe, of data frames)
```

The benefit here is consistency in names and data structure; both remain well known. The downside is in loss of control over the structure of the `goals` and `simulated_goals` data frames.  Both would have to have some minimum number of columns, but a user out to be be able to add columns to ensure that they can save what the user deems  necessary for an accelerated goals object (note that doing so means we require the user to use dplyr's `rbind_all` or `rbind_list`

Another difficulty is mixing different types of goals (say a one iteration goals vs goals projected into the future).  Different types of goals may require different "by"/key columns when joining goal data back to the `growth_df`; how would we allow the user to identify theses columns?

**Comment by chrishaid** [comment]
A third approach is to combine the two:  allow stacking into a well defined date.frame _and_ maintain custom objects in custom named lists. Literally combine the two approaches above.

**Comment by chrishaid** [comment]
Fourth idea:  Use the original design and always populate a canonical set of `goals` objects (typical, KIPP Tiered, 75th percentile of growth, Rutgers Ready, and simulation. 

That way we write visualizations to those standard goal types with well known names. Those visualizations should have an optional argument for passing the the name of the goals object the user wants.  Further, I think any user that has the wherewithal to write their own `goals_function` will understand what's happening, especially give the well crafted RMD I promise to write! :dizzy: :grinning: 

@almartin82 Thoughts?

**Comment by chrishaid** [comment]
Another idea that is a meeting in the middle two slots for `goals` objects. A canonical goals slot, with what we deem important in there, and a custom slot for user created `goals objects`.  So something like this:

```
mapvizieR_object
|----cdf
|----roster
|----growth_df
|----goals (canonical goals objects go here =)
      |----typcial_growth 
            |----goals 
            |----join_by_fields 
            |----slot_name 
      |----kipp_tiered_growth 
            |----goals 
            |----join_by_fields 
            |----slot_name 
      |----rutgers_ready_growth 
            |----goals
            |----join_by_fields 
            |----slot_name
|----custom_goals (this slot only created if a custom goal object is created).
      |----someones_super_cool_simulated_annealing_goals
            |----goals
            |----join_by_fields 
            |----slot_name
      |----anothers_radical_EM_algorithm_goals
            |----goals
            |----join_by_fields 
            |----slot_name 
```

I like this one the best.  It is relatively flexible, and if we find that a simple data.frame will suffice it is easy to port this code from lists to dfs and to change functions to work well with this.  If on the other hand, we find the data.frame approach too limiting, then porting to lists objects will probably require a bit more work.

@almartin82, so why don't we merge this puppy as it is. and we can work understanding goal setting in a bit, but focus on viz's for now. Pretty please. :cherries: :arrow_double_up:

**Comment by almartin82** [comment]
I really like this last compromise structure.  If it's not that different, keep it in the goals data frame with a name; put other stuff into custom goals.

**Comment by almartin82** [comment]
Ok from our call we're going to merge this now, and open issues to move the goals stuff over to two slots: goals and custom goals.

**Pull request merged**